cpp/validate_harvested_records.cc

```diff
@@ -39,274 +40,195 @@
-    const_iterator find(const std::string &field_name) const {
-        return std::find_if(field_infos_.begin(), field_infos_.end(),
-                            [&field_name](const FieldInfo &field_info){ return field_name == field_info.name_; });
-    }
-    iterator begin() { return field_infos_.begin(); }
-    iterator end() { return field_infos_.end(); }
-    iterator find(const std::string &field_name) {
-        return std::find_if(field_infos_.begin(), field_infos_.end(),
-                            [&field_name](const FieldInfo &field_info){ return field_name == field_info.name_; });
-    }
+void FieldRule::addRule(const char subfield_code, const FieldPresence field_presence) {
+    if (unlikely(subfield_code_to_field_presence_map_.find(subfield_code) != subfield_code_to_field_presence_map_.end()))
+        LOG_ERROR("Attempt to insert a second rule for subfield code '" + std::string(1, subfield_code) + "'!");
+    subfield_code_to_field_presence_map_[subfield_code] = field_presence;
+}
 
-    // Combine GeneralInfo with other General Info (e.g. JournalInfo).
-    // rhs will have priority to simulate data inheritance.
-    static GeneralInfo Combine(const GeneralInfo &lhs, const GeneralInfo &rhs);
-};
 
+void FieldRule::findRuleViolations(const MARC::Subfields &subfields, std::string * const reason_for_being_invalid) const {
+    std::set<char> found_subfield_codes;
+    for (const auto &subfield : subfields)
+        found_subfield_codes.emplace(subfield.code_);
 
-GeneralInfo GeneralInfo::Combine(const GeneralInfo &lhs, const GeneralInfo &rhs) {
-    auto lhs_iter(lhs.begin());
-    auto rhs_iter(rhs.begin());
-
-    GeneralInfo combined_info;
-    while (lhs_iter != lhs.end() or rhs_iter != rhs.end()) {
-        if (lhs_iter == lhs.end()) {
-            combined_info.addField(*rhs_iter);
-            ++rhs_iter;
-        } else if (rhs_iter == rhs.end()) {
-            combined_info.addField(*lhs_iter);
-            ++lhs_iter;
-        } else {
-            const int compare_result(StringUtil::AlphaWordCompare(lhs_iter->name_, rhs_iter->name_));
-            if (compare_result == 0) {
-                // if present on both sides, rhs wins!
-                combined_info.addField(*rhs_iter);
-                ++lhs_iter;
-                ++rhs_iter;
-            } else if (compare_result < 0) {
-                combined_info.addField(*lhs_iter);
-                ++lhs_iter;
-            } else if (compare_result > 0) {
-                combined_info.addField(*rhs_iter);
-                ++rhs_iter;
-            }
+    for (const auto &subfield_code_and_field_presence : subfield_code_to_field_presence_map_) {
+        const auto iter(found_subfield_codes.find(subfield_code_and_field_presence.first));
+        if (iter == found_subfield_codes.end() and subfield_code_and_field_presence.second == ALWAYS) {
+            if (not reason_for_being_invalid->empty())
+                reason_for_being_invalid->append("; ");
```

Comment:
Personal taste: i would rather use a `std::vector<std::string>` for `reason_for_being_invalid` and leave the way things are formatted for output to the caller (e.g. to implode it with `;`, `\n` or anything else)